torchtune/data/_prompt_templates.py

```diff
@@ -0,0 +1,154 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+from functools import partial
+from typing import Dict, List, Protocol, Tuple
+
+from torchtune.data import Message, Role
+
+
+class PromptTemplate(Protocol):
+    """
+    Interface for prompt templates. Each prompt template can include structured
+    text for system, user, and assistant roles that are prepended or appended to
+    the message content.
+    """
+
+    # Template should map role to a tuple containing the tag to prepend to the text
+    # and tag to append to the text. Leave as empty strings to not prepend or append
+    template: Dict[Role, Tuple[str, str]]
+
+    def __call__(
+        self,
+        messages: List[Message],
+    ) -> List[Message]:
+        """
+        Format each role's message(s) according to the prompt template
+
+        Args:
+            messages (List[Message]): a single conversation, structured as a list
+                of :class:`~torchtune.data.Message` objects
+
+        Returns:
+            The formatted list of messages
+        """
+        pass
+
+
+class CustomPromptTemplate(PromptTemplate):
```

Comment:
I'm in favor of renaming this to `PromptTemplate` and call the other thing `PromptTemplateInterface`.

The "Custom" part of prompt template feels redundant. Like what's the difference between a CustomPromptTemplate and a regular PromptTemplate? I see here by looking at the code that it's b/c one is an interface, but that's not clear from the names. This should be evident before having to go to the docs.

Reply:
We also use the Interface naming for protocols with our recipes so it makes sense.

Reply:
that makes sense to me, although if a user needs a bit more custom behavior that's not offered by `CustomPromptTemplate`, I wouldn't want them to accidentally inherit from `PromptTemplate` instead of `PromptTemplateInterface`.